dot/rpc/http_test.go

```diff
@@ -90,3 +100,256 @@ func TestNewHTTPServer(t *testing.T) {
+		RPCPort:    7880,
+		RPCAPI:     NewService(),
+		RPCUnsafe:  true,
+		NetworkAPI: netmock,
+	}
+
+	s := NewHTTPServer(cfg)
+	err = s.Start()
+	require.NoError(t, err)
+
+	time.Sleep(time.Second)
+
+	ip, err := externalIP()
+	require.NoError(t, err)
+
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("http://%s:7880/", ip), buf)
+	require.NoError(t, err)
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := new(http.Client).Do(req)
+	require.NoError(t, err)
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	require.NoError(t, err)
+
+	expected := `{"jsonrpc":"2.0","error":{"code":-32000,"message":"external HTTP request refused","data":null},"id":1}` + "\n"
+	require.Equal(t, expected, string(resBody))
+}
+
+func TestRPCExternalEnable_UnsafeExternalNotEnabled(t *testing.T) {
+	unsafeMethod := "system_addReservedPeer"
+	unsafeData := []byte(fmt.Sprintf(
+		`{"jsonrpc":"2.0","method":"%s","params":["%s"],"id":1}`,
+		unsafeMethod,
+		"/ip4/198.51.100.19/tcp/30333/p2p/QmSk5HQbn6LhUwDiNMseVUjuRYhEtYj4aUZ6WfWoGURpdV"))
+	unsafebuf := new(bytes.Buffer)
+	unsafebuf.Write(unsafeData)
+
+	safeMethod := "system_localPeerId"
+	safeData := []byte(fmt.Sprintf(
+		`{"jsonrpc":"2.0","method":"%s","params":[],"id":2}`,
+		safeMethod))
+	safebuf := new(bytes.Buffer)
+	safebuf.Write(safeData)
+
+	netmock := new(mocks.MockNetworkAPI)
+	netmock.On("NetworkState").Return(common.NetworkState{
+		PeerID: "peer id",
+	})
+
+	cfg := &HTTPServerConfig{
+		Modules:     []string{"system"},
+		RPCPort:     7880,
+		RPCAPI:      NewService(),
+		RPCUnsafe:   true,
+		RPCExternal: true,
+		NetworkAPI:  netmock,
+	}
+
+	s := NewHTTPServer(cfg)
+	err := s.Start()
+	require.NoError(t, err)
+
+	time.Sleep(time.Second)
+
+	ip, err := externalIP()
+	require.NoError(t, err)
+
+	// safe method should be ok
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("http://%s:7880/", ip), safebuf)
+	require.NoError(t, err)
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err := new(http.Client).Do(req)
+	require.NoError(t, err)
+
+	resBody, err := ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	require.NoError(t, err)
+
+	encoded := base58.Encode([]byte("peer id"))
+	expected := fmt.Sprintf(`{"jsonrpc":"2.0","result":"%s","id":2}`, encoded) + "\n"
+	require.Equal(t, expected, string(resBody))
+
+	// unsafe method should not be ok
+	req, err = http.NewRequest(http.MethodPost, fmt.Sprintf("http://%s:7880/", ip), unsafebuf)
+	require.NoError(t, err)
+	req.Header.Set("Content-Type", "application/json")
+
+	res, err = new(http.Client).Do(req)
+	require.NoError(t, err)
+
+	resBody, err = ioutil.ReadAll(res.Body)
+	defer res.Body.Close()
+	require.NoError(t, err)
+
+	expected = `{"jsonrpc":"2.0","error":{"code":-32000,"message":"external HTTP request refused","data":null},"id":1}` + "\n"
+	require.Equal(t, expected, string(resBody))
+}
+
+func makePOSTRequest(t *testing.T, url string, data io.Reader) (int, []byte) {
```

Comment:
use the variable `url`
Currently, the value is hardcoded. 
```
http://localhost:7878/
```

Reply:
Fixed!

Reply:
done